core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -345,13 +345,29 @@ private static RexNode canonizeNode(RexBuilder rexBuilder, RexNode condition) {
     case LESS_THAN_OR_EQUAL:
     case GREATER_THAN_OR_EQUAL: {
       RexCall call = (RexCall) condition;
-      final RexNode left = call.getOperands().get(0);
-      final RexNode right = call.getOperands().get(1);
+      RexNode left = call.getOperands().get(0);
+      left = canonizeNode(rexBuilder, left);
+      RexNode right = call.getOperands().get(1);
+      right = canonizeNode(rexBuilder, right);
+      call = (RexCall) rexBuilder.makeCall(call.getOperator(), left, right);
       if (left.toString().compareTo(right.toString()) <= 0) {
         return call;
```

Comment:
You can only return the original call if `left==call.getOperands().get(0) && right==call.getOperands().get(1)`. Below (for + and *), you could check if that is the case too, and return the original call otherwise; it is worth skipping new expression creation in these cases).